src/metabase/channel/render/body.clj

```diff
@@ -388,11 +389,138 @@
+         :funnel.dimension 'DIMENSION',
+         :funnel.order_dimension 'DIMENSION',
+         ...}}}
+
+   The input `series-data` is the dashcard data series results from QP as a vector of maps, [{:card {...} :data {...}, ...]
+  "
+  [series-data {:keys [columns columnValuesMapping] :as visualizer-settings}]
+  (let [source-mappings-with-vals   (extract-value-sources columnValuesMapping)
+        ;; Create map from virtual column name e.g. 'COLUMN_1' to a vector of values only for value sources
+        remapped-col-name->vals     (reduce
+                                     (fn [acc {:keys [name originalName] :as source-mapping}]
+                                       (let [ref-card-id      (value-source->card-id source-mapping)
+                                             card-with-data   (u/find-first-map series-data [:card :id] ref-card-id)
+                                             card-cols        (get-in card-with-data [:data :cols])
+                                             card-rows        (get-in card-with-data [:data :rows])
+                                             col-idx-in-card  (u/find-first-map card-cols :name originalName true)]
+                                         (if col-idx-in-card
+                                           (let [values (mapv #(nth % col-idx-in-card) card-rows)]
+                                             (assoc acc name values))
+                                           acc)))
+                                     {}
+                                     source-mappings-with-vals)
+        ;; Create column-major matrix for all virtual columns, value and name ref sources
+        unzipped-rows               (mapv
+                                     (fn [column]
+                                       (let [source-mappings (get columnValuesMapping (keyword (:name column)))]
+                                         (->> source-mappings
+                                              (mapcat
+                                               (fn [source-mapping]
+                                                 (if (source-is-name-ref? source-mapping)
+                                                   ;; Source is a name ref so just return the name of the card with matching :id
+                                                   (let [card-id (name-source->card-id source-mapping)
+                                                         card    (:card (u/find-first-map series-data [:card :id] card-id))]
+                                                     (if-let [name (:name card)]
+                                                       [name]
+                                                       []))
+                                                   ;; Source is actual column data
+                                                   (let [values (get remapped-col-name->vals (:name source-mapping))]
+                                                     (if values values [])))))
```

Comment:
```suggestion
                                                     (or values [])))))
```

Reply:
or maybe just `values` works here? if you return `nil` it'll just be skipped by `mapcat`